### PR TITLE
Shared wallets in incomplete state should not undergo restoration

### PR DESCRIPTION
**Pull request by paweljakubas**
### Issue Number

ADP-2432

### Description

- [x] I have made sure `Incomplete` shared wallet invokes  `createNonRestoringWalletWorker`.
- [x] When `Incomplete -> Ready` transition then the private key and meta from incomplete wallet is read.
- [x] The incomplete wallet is deleted and active wallet is created. The wallet starts restoring. 
- [x] Wallet takes meta and private key from incomplete wallet.

The solution making db exchange between workers (non-restoring and restoring) did not work properly. hence the above solution.

### CODE REVIEW

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs [hunk 1]
```diff
@@ -769,6 +769,11 @@ instance IsServerError ErrDerivePublicKey where
 instance IsServerError ErrAddCosignerKey where
     toServerError = \case
         ErrAddCosignerKeyNoSuchWallet e -> toServerError e
+        ErrAddCosignerKeyNoMeta ->
+            apiError err503 NoMeta $ mconcat
+                [ "No meta of wallet was found in database during migration"
```
Comment (jonathanknowles) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs]: @paweljakubas 

When you refer to "no meta of wallet", do you mean "no wallet metadata"?

**Review by jonathanknowles — COMMENTED** [review]
Hi @paweljakubas 

Thanks for making this PR!

I've given it a brief look (though still need to look further).

I have one question about the use of terminology:

When you say **"no meta"**, are you referring to a case where the database has **no metadata** for a given wallet?

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs (hunk 1 above)
Comment (paweljakubas) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs]: yes

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs [hunk 2]
```diff
@@ -769,10 +769,12 @@ instance IsServerError ErrDerivePublicKey where
...
         ErrAddCosignerKeyNoSuchWallet e -> toServerError e
-        ErrAddCosignerKeyNoMeta ->
-            apiError err503 NoMeta $ mconcat
-                [ "No meta of wallet was found in database during migration"
-                , " from pending to active state. Could be database error."
+        ErrAddCosignerKeyWalletMetadataNotFound ->
+            apiError err503 WalletMetadataNotFound $ T.unwords
+                [ "It was not possible to find any metadata for the given"
+                , "wallet within the database. This could be because the"
+                , "wallet has yet to become active after being in the pending"
```
Comment (piotr-iohk) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs]:
```suggestion
                , "wallet has yet to become active after being in the incomplete"
```

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs [hunk 3]
```diff
@@ -1156,13 +1193,56 @@ patchSharedWallet
     -> ApiSharedWalletPatchData
     -> Handler ApiSharedWallet
 patchSharedWallet ctx liftKey cred (ApiT wid) body = do
+    wal <- fst <$> getWallet ctx (mkSharedWallet @_ @s @k) (ApiT wid)
     withWorkerCtx ctx wid liftE liftE $ \wrk -> do
         liftHandler $ W.updateCosigner wrk wid (liftKey accXPub) cosigner cred
+    wal' <- fst <$> getWallet ctx (mkSharedWallet @_ @s @k) (ApiT wid)
+    -- we switch on restoring only after pending -> active transition
```
Comment (piotr-iohk) [inline lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]:
```suggestion
    -- we switch on restoring only after incomplete -> active transition
```

File: specifications/api/swagger.yaml [hunk 4]
```diff
@@ -4165,6 +4165,22 @@ x-errNoRootKey: &errNoRootKey
...
+  title: wallet_metadata_not_found
+  properties:
+    message:
+      type: string
+      description: |
+        Indicates that it was not possible to find any metadata for the given
+        wallet within the database.
+
+        May occur when a shared wallet has not yet become active after being in
+        the pending state.
```
Comment (piotr-iohk) [inline specifications/api/swagger.yaml]:
```suggestion
        the incomplete state.
```

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs (hunk 2 above)
Comment (piotr-iohk) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs]: Q: what are circumstances for this error to occur? Is it possible to produce the test that incurs this error?

**Review by piotr-iohk — COMMENTED** [review]
I have played a bit and it seems to work as expected. Also e2e tests pass on the branch.

I think that we could have some integration tests:
 - producing the introduced `WalletMetadataNotFound` if that is possible
 - creating a shared wallet, funding it, then deleting it and creating again "in steps", i.e. create + patch and making sure the balance is as expected.

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs (hunk 2 above)
Comment (paweljakubas) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Server/Error.hs]: if something wrong with db ie. when during being in incomplete db is erased

File: lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs (hunk 3 above)
Comment (paweljakubas) [reply lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs]: done

File: specifications/api/swagger.yaml (hunk 4 above)
Comment (paweljakubas) [reply specifications/api/swagger.yaml]: done

**Comment by paweljakubas** [comment]
Two things added:
1. Made pending -> incomplete replacement in whole code base
2. Added integration test showing the case that Piotr suggested
@piotr-iohk @jonathanknowles

File: lib/wallet/integration/src/Test/Integration/Scenario/API/Shared/Transactions.hs [hunk 5]
```diff
@@ -837,7 +909,74 @@ spec = describe "SHARED_TRANSACTIONS" $ do
...
+        let walPending2 = getFromResponse Prelude.id rPost2
+        let payloadPatch2 = Json [json| {
+                "cosigner#0": #{accXPubDerived1}
+                } |]
+        rPatch2 <- patchSharedWallet ctx walPending2 Payment payloadPatch2
+        expectResponseCode HTTP.status200 rPatch2
+        let walShared2 = getFromResponse Prelude.id rPatch2
+
+        --checking that balances are the same as before deletion
+        eventually "balance is like before deletion" $ do
```
Comment (piotr-iohk) [inline lib/wallet/integration/src/Test/Integration/Scenario/API/Shared/Transactions.hs:967]: :+1:

File: specifications/api/swagger.yaml [hunk 6]
```diff
@@ -5208,6 +5224,13 @@ x-responsesErr404WalletNotFound: &responsesErr404WalletNotFound
       application/json:
         schema: *errNoSuchWallet
 
+x-responsesErr503WalletMetadataNotFound: &responsesErr503WalletMetadataNotFound
+  503:
+    description: No meta
```
Comment (piotr-iohk) [inline specifications/api/swagger.yaml:5229]: A bit more elaborate description would be nice. :)

File: test/e2e/spec/spec_helper.rb [hunk 7]
```diff
@@ -175,7 +175,7 @@ def patch_if_incomplete(wid, payment_patch, deleg_patch)
     expect(d_upd).to be_correct_and_respond 200
   when 403
     expect(d_upd).to be_correct_and_respond 403
-    expect(d_upd.parsed_response['code']).to eq 'shared_wallet_not_pending'
+    expect(d_upd.parsed_response['code']).to eq 'shared_wallet_active'
```
Comment (piotr-iohk) [inline test/e2e/spec/spec_helper.rb:178]: :+1:

**Review by piotr-iohk — APPROVED** [review]
LGTM.

**Comment by Anviking** [comment]
bors r-

_restarting_

**Pull request merged**